src/main/java/io/snowdrop/github/issues/GithubIssueBridge.java

```diff
@@ -0,0 +1,296 @@
+package io.snowdrop.github.issues;
+
+import java.io.IOException;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.egit.github.core.Issue;
+import org.eclipse.egit.github.core.Label;
+import org.eclipse.egit.github.core.User;
+import org.eclipse.egit.github.core.client.GitHubClient;
+import org.eclipse.egit.github.core.client.GitHubRequest;
+import org.eclipse.egit.github.core.client.IGitHubConstants;
+import org.eclipse.egit.github.core.service.IssueService;
+import org.eclipse.egit.github.core.service.LabelService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.snowdrop.BotException;
+import io.snowdrop.github.Github;
+
+public class GithubIssueBridge {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(GithubIssueBridge.class);
+
+  public static final String SEPARATOR = "\n\n---\n\n";
+  public static final String UPSTREAM_REPO_PREFIX = "$upstream:";
+  public static final String UPSTREAM_REPO_SUFFIX = "$";
+
+  public static final String ASSIGNEES = "assignees";
+
+  private final GitHubClient client;
+  private final IssueService issueService;
+  private final LabelService labelService;
+
+  private final String sourceRepository;
+  private final String targetRepository;
+  private final String terminalLabel;
+  private final Set<String> users;
+
+  private final Map<String, Map<String, Label>> repoLables = new HashMap<>();
+
+  private final Map<Integer, Issue> openIssues = new HashMap<>();
+  private final Map<Integer, Issue> closedIssues = new HashMap<>();
+  private final Map<Integer, Issue> downstreamIssues = new HashMap<>();
+
+  public GithubIssueBridge(GitHubClient client, IssueService issueService, LabelService labelService, String sourceRepository, String targetRepository, String terminalLabel, Set<String> users) {
+    this.client = client;
+    this.issueService = issueService;
+    this.labelService = labelService;
+    this.sourceRepository = sourceRepository;
+    this.targetRepository = targetRepository;
+    this.terminalLabel = terminalLabel;
+    this.users = users;
+  }
+
+  public synchronized void init() {
```

Comment:
Why are all these methods `synchronized`

Reply:
The client doesn't support multi-thread access (it will barf), so all entrypoints to the client need to be protected.